README
=head1 NAME

README - Perl Core Test Smoke Suite (1.30)

=head1 SYNOPSIS

  $ perl configsmoke.pl
  $ sh smokecurrent.sh

=head1 DESCRIPTION

The perl core test smoke suite is a set of scripts and modules that try to
run the perl core tests on as many configurations as possible and combine
the results into an easy to read report.

The basic cycle looks like:

    fetch_sourcetree
    foreach $config ( @configurations ) {
        make distclean
        ./Configure $config
        make
        make test (and log failures)
    }
    create_report()
    mail_report()

Before you start "smoking" we expect you have tried to build some of
the configurations "by hand" to check if your environment is okay.

=head1 INSTALLATION

The Test::Smoke suite has been designed to be installed outside of the
normal Perl library-tree. It will also run from the directory in which
it was un-tar-gzipped, but if you like to keep up to date with
Test::Smoke releases, it might be useful to use the standard
perl-module installation commands:

    $ perl Makefile.PL
    $ make test
    $ make install

The F<Makefile.PL> program will ask for a directory in which
Test::Smoke should be installed (this will set I<PREFIX>,
I<INSTALLSCRIPT> and I<LIB> all to that directory).

I<If you want to automate the Makefile.PL step you need to set two
environment variables: $ENV{SMOKE_INSTDIR}='../p5smoke' and
$ENV{PERL_MM_USE_DEFAULT}='y'>

You are now ready to run the F<configsmoke.pl> configuration program:

    $ perl configsmoke.pl

You will have to answer some questions that help create a
configuration file for the Test::Smoke suite.

You should check the F<perlcurrent.cfg> build-configurations file and
remove options that are not supported for your platform.

Now you are set to go:

    $ sh smokecurrent.sh

If something useful happens, do it every night using a scheduler.

=head1 CONFIGURATION

=over 4

=item Test::Smoke configuration

You can create configurations by running the F<configsmoke.pl>
program. Please see its L<documentation|configsmoke.pl>.

=item Build configurations

At the moment there are no guidelines of how to change the
build-configuration best to fit the needs of perl5-porters while still
getting as much smoke out of the test as possible, simply because the
needs of perl5-porters change over time:).

This also makes Test::Smoke somewhat volatile, as we aim to please the
perl5-porters as much as possible, to help make perl even better, more
stable and as thoroughly tested as possible.

=back

=head1 Note to MSWin32 users

To make it clear, the perl core test smoke suite treats B<cygwin> and
B<DJGPP> as unixy platforms and not as true MSWin32.

To keep in sync with the source-tree, it is recommended you use the
B<rsync> program. This currently seems only available with Cygwin
<http://www.cygwin.com>. (All the cygwin tools run perfectly well
under CMD.EXE, so don't let that stop you.)

Some problems have been reported with older versions of cygwin and the
patch program during the C<< S<[nd]make test> >> stage of
Test::Smoke. Don't worry about it if you do not intend to use patch.

Windows 95/98/ME do not seem to be ideal platforms to run Test::Smoke
on. F<COMMAND.COM> gives some problems while building perl. (See
F<README.win32>.)

We have found some problems with the default windows debugger
B<Dr.Watson>. By default Dr.Watson will pop-up when the moral equivalent
of a segmentation fault happens, this will make the smoke wait for
user input. You can disable this by starting F<drwtsn32.exe> and
un-check the B<Visual Notification> and B<Sound Notification>
check-boxes and press B<OK>.

There seems to be a problem with B<Microsoft .NET/Windows XP> and
segmentation faults, even if B<drwtsn32 -i> is run to set it as the
default debugger.

=head1 Slow Systems

It can happen that your system is not able to pursue all the tests, either
because it is too slow or because some of the configurations are not
supported by the system. At first, don't worry, they are detected by the
report creation script and shown as such.

After the first run gave you some idea of how long the smoke will run on
your system,  you can create your own copy of smoke.cfg where you either
add new combinations and/or test levels, or delete lines to speed up the
run time of Test::Smoke.

=head1 Availability and references

=over 4

=item B<Distribution>

    http://search.cpan.org/dist/Test-Smoke/

=item B<Archives>

=over 4

=item B<Reports>

    <daily-build-reports@perl.org> 
        [subscribe: <daily-build-reports-subscribe@perl.org>]
    http://www.nntp.perl.org/group/perl.daily-build.reports
    news:nntp.perl.org/perl.daily-build.reports

=item B<Smokers>

    <daily-build@perl.org> 
        [subscribe: <daily-build-subscribe@perl.org>]
    http://www.nntp.perl.org/group/perl.daily-build
    news:nntp.perl.org/perl.daily-build

=item B<Others>

  Perl general QA: perl-qa@perl.org
  Development:     perl5-porters@perl.org

=item B<The new smoke_db> (by Alain Barbet)

Alain Barbet has put a lot of effort in storing the smoke test
reports in a database and build a nice WEB-front-end:

    http://www.alianwebserver.com/cgi-bin/smoke_db

=item B<TinySmoke db>

I also keep a database of smokereports that are created by Test::Smoke
1.17 or higher.

    http://www.test-smoke.org/cgi/tsdb

=back

=back

=head1 REVISION

$Id: README 1155 2008-01-03 13:32:28Z abeltje $

=head1 COPYRIGHT and LICENSE

Copyright (C) 2001-2004

  * H.Merijn Brand
  * Abe Timmerman

This suite is free software; you can redistribute it and/or modify it
under the same terms as Perl itself, without consulting the author.

(Future) Co-Authors and or contributors should agree to this before
submitting patches.

See:

=over 4

=item * http://www.perl.com/perl/misc/Artistic.html

=item * http://www.gnu.org/copyleft/gpl.html

=back

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


=head1 AUTHOR

The perl core test smoke suite is maintained by Abe Timmerman
<abeltje@cpan.org>.

H.Merijn Brand <h.m.brand@hccnet.nl> in dialogue and consultation with
Michael Schwern <schwern@pobox.com>, the Perl QA pumpkin.

Contributors and/or co-authors:

    Configurations and
      nuts and bolts	Nicholas Clark <nick@plum.flirble.org>
    smoke.sh            Jarkko Hietaniemi <jhi@cc.hut.fi>
    Portability issues	Will Coleda - IMG <wjc@infomg.com>
    Safety		Richard Soderberg <rs@oregonnet.com>
    Win32/MSC5/dmake	Mattia Barbon <mbarbon@dsi.unive.it>,
			Abe Timmerman <abeltje@cpan.org>
    Cygwin		John Peacock <jpeacock@rowman.com>
    Sharing issues	Blair Zajac <blair@orcaware.com>
    Good ideas          Jos Boumans <kane@dwim.org>
    VMS                 Craig Berry
    SysInfo             Alain Barbet
                        Merijn Brand
                        Alan Burlison
                        Allen Smith
                        David Woolger
                        Campo Wijerman

and all the people having useful suggestions and nits, starting at
YAPC::Europe-2.0.01, where Michael talked Merijn into simplifying his own
test-suite and sharing it with the community.

The Test::Smoke team.

=head1 MISC

As always, have the appropriate amount of fun!

=head1 SEE ALSO

L<ReleaseNotes>, L<FAQ>

=head1 TODO

=over 4

=item * More test-suite

=item * VMS?

=back

=cut


Changes
These changes are from the Subversion repository at
        Url: https://source.test-smoke.org/svn/branches/Test-Smoke-1.30.xx

The filelists have a short symbolic indicator:
        !  modified
        +  added
        -  deleted
       !>  merged change
       !<  replaced

Enjoy!
___________________________________________________________________________
[  1284] By: abeltje                          on 2010-11-01 19:12:17 +0100
        Log: * [MERGE]
	     [	1280] By: tux		    on 2010-08-08 13:44:20 +0200
		     Log: Add user_note attribute
        Url: https://source.test-smoke.org/svn/branches/Test-Smoke-1.30.xx
           ! /branches/Test-Smoke-1.30.xx,
	     /branches/Test-Smoke-1.30.xx/lib/Test/Smoke/Reporter.pm,
	     /branches/Test-Smoke-1.30.xx/lib/Test/Smoke/SysInfo.pm,
	     /branches/Test-Smoke-1.30.xx/t/reporter.t
___________________________________________________________________________
[  1271] By: abeltje                          on 2010-02-12 22:44:04 +0100
        Log: * [MERGE]
	     [	1270] By: bram		     on 2010-01-21 08:59:39 +0100
		     Log: [FIX] Update MANIFEST with new files
	     * [CHANGE]
	       + Up the version.
        Url: https://source.test-smoke.org/svn/branches/Test-Smoke-1.30.xx
           ! /branches/Test-Smoke-1.30.xx,
	     /branches/Test-Smoke-1.30.xx/MANIFEST,
	     /branches/Test-Smoke-1.30.xx/lib/Test/Smoke.pm
___________________________________________________________________________
[  1268] By: abeltje                          on 2010-01-20 22:31:28 +0100
        Log: * [AUTOCOMMIT]
	       * Regenerate 'Changes'
        Url: https://source.test-smoke.org/svn/branches/Test-Smoke-1.30.xx
           ! /branches/Test-Smoke-1.30.xx/Changes
___________________________________________________________________________
[  1267] By: abeltje                          on 2010-01-20 22:29:58 +0100
        Log: * [MERGE] from ts-current
	     [	1254] By: bram				   on 2009-11-16
	     19:03:04 +0100
		     Log: Fixes:

			  - String after 'Finished Smoking' can also
			    contain [a-zA-Z]
			    (git id)
			  - Fix write_to_file (created the file 1 instead
			    of the actual
			    name)
			  - Fix error reporting in set_skip_tests: if a
			    file was
			    succesfully renamed it reported 'not ok'
			    instead of 'ok' +
			    error message was not shown
			  - Fix set_smoke_patchlevel: the smoke id can
			    contain [a-zA-Z]
	     [	1255] By: bram		       on 2009-11-20 00:25:01
	     +0100
	       Log: Add tests for earlier commit.

		    In the reporter test 3 different patchlevels:
		    - all numeric (perforce patch number)
		    - hex numbers: starting with a number (git commit id)
		    - hex numbers: starting with a letter (git commit id)
	     [	1256] By: bram		       on 2009-11-21 14:03:46
	     +0100
	       Log: [FIX]
			Tests that are skipped in dist/ and cpan/ have to
			be removed from MANIFEST
	     [	1257] By: bram		       on 2009-11-21 14:17:20
	     +0100
	       Log: Forgot test for previous fix
	     [	1258] By: bram		       on 2010-01-10 19:33:44
	     +0100
	       Log: Add 'Passed Todo tests' section in the report
	     [	1259] By: bram		       on 2010-01-11 21:49:31
	     +0100
	       Log: - Include 'Non-zero exit status' in the output file
		    - Include 'Non-zero exit status' in the report
		    - If the harness output is not parsed for a test file
		      then add	'test_file...??????'  in the failed list
	     [	1261] By: abeltje	       on 2010-01-12 21:58:28
	     +0100
	       Log: * [CHANGE]
		      * Merge my changes with Bram's
	     [	1263] By: bram		       on 2010-01-12 23:13:19
	     +0100
	       Log: Fix failing tests.
		    The output now includes 'Non-zero exit status'
	     [	1264] By: abeltje	       on 2010-01-12 23:36:27
	     +0100
	       Log: * [FIX] thanks to Bram
		     * the new regex for finding the smoke-git-id
		       shouldn't test for leading TAB, these are stripped
		       by Test::Smoke::Util::get_local_patches().
	     [	1265] By: bram		       on 2010-01-12 23:39:48
	     +0100
	       Log: $last and next is not needed. (breaks the parsing of
	       passed todo tests)
	     [	1266] By: bram		       on 2010-01-13 11:43:18
	     +0100
	       Log: Set $! to 0 at the start of the test
        Url: https://source.test-smoke.org/svn/branches/Test-Smoke-1.30.xx
           ! /branches/Test-Smoke-1.30.xx,
	     /branches/Test-Smoke-1.30.xx/lib/Test/Smoke/Reporter.pm,
	     /branches/Test-Smoke-1.30.xx/lib/Test/Smoke/Smoker.pm,
	     /branches/Test-Smoke-1.30.xx/lib/Test/Smoke.pm,
	     /branches/Test-Smoke-1.30.xx/mkdist.sh,
	     /branches/Test-Smoke-1.30.xx/private/perl/t/smoke/die.t,
	     /branches/Test-Smoke-1.30.xx/private/perl/t/smoke/many.t,
	     /branches/Test-Smoke-1.30.xx/private/perl/t/smoke/minitest.t,
	     /branches/Test-Smoke-1.30.xx/private/perl/t/smoke/test.t,
	     /branches/Test-Smoke-1.30.xx/private/smoker_fail.t,
	     /branches/Test-Smoke-1.30.xx/t/ftppub/bugtst01.out,
	     /branches/Test-Smoke-1.30.xx/t/ftppub/bugtst02.out,
	     /branches/Test-Smoke-1.30.xx/t/ftppub/bugtst03.out,
	     /branches/Test-Smoke-1.30.xx/t/ftppub/perl-current/MANIFEST,
	     /branches/Test-Smoke-1.30.xx/t/reporter.t,
	     /branches/Test-Smoke-1.30.xx/t/smoker.t
          !< /branches/Test-Smoke-1.30.xx/t/ftppub/perl-current/cpan/t
	     =(/Test-Smoke/t/ftppub/perl-current/cpan/t:1257),
	     /branches/Test-Smoke-1.30.xx/t/ftppub/perl-current/cpan/t/ski
	     p.t =(/Test-Smoke/t/ftppub/perl-current/cpan/t/skip.t:1257),
	     /branches/Test-Smoke-1.30.xx/t/ftppub/perl-current/cpan/t/tes
	     t.t =(/Test-Smoke/t/ftppub/perl-current/cpan/t/test.t:1257),
	     /branches/Test-Smoke-1.30.xx/t/ftppub/perl-current/dist/t
	     =(/Test-Smoke/t/ftppub/perl-current/dist/t:1257),
	     /branches/Test-Smoke-1.30.xx/t/ftppub/perl-current/dist/t/ski
	     p.t =(/Test-Smoke/t/ftppub/perl-current/dist/t/skip.t:1257),
	     /branches/Test-Smoke-1.30.xx/t/ftppub/perl-current/dist/t/tes

Makefile.PL
#! /usr/bin/perl -w
# $Id: Makefile.PL 1155 2008-01-03 13:32:28Z abeltje $
use strict;
use 5.004;
use ExtUtils::MakeMaker;
use Config;

my $findbin;
use File::Basename;
BEGIN { $findbin = dirname $0 }
use File::Spec::Functions qw( :DEFAULT rel2abs );
use Cwd;
use File::Path;

# Some strange stuff happens when WriteMakefile() is called
# on a non-clean distribution, cleanup before
MAKE_DISTCLEAN: {
    local *SAVEERR; open SAVEERR, ">&STDERR";
    local *SAVEOUT; open SAVEOUT, ">&STDOUT";
    close STDERR; close STDOUT;

    my $opt = $^O eq 'VMS' ? '/IGNORE' : ' -i';
    system "$Config{make}$opt distclean";

    open STDOUT, ">&SAVEOUT"; close SAVEOUT;
    open STDERR, ">&SAVEERR"; close SAVEERR;
}

my $mmver = ExtUtils::MakeMaker->VERSION;

my $iscript = 'INSTALLSITESCRIPT';
unless ( eval "$mmver" >= eval "6.30_01" ) {
    print "You are running MakeMaker v$mmver: fix INSTALLSCRIPT\n";
    $iscript = 'INSTALLSCRIPT';
}

my %install = ( );
unless ( @ARGV ) {
    my $dft_install = exists $ENV{SMOKE_INSTDIR} && $ENV{SMOKE_INSTDIR} ne ""
        ? $ENV{SMOKE_INSTDIR}
        : catdir( updir(), 'smoke' );
    $dft_install = rel2abs( $dft_install );

    my $install_dir = dir_prompt( 
        "Where would you like to install Test::Smoke?\n",
        $dft_install
    );

    %install = ( PREFIX => $install_dir, $iscript => $install_dir,
#                 LIB   => File::Spec->catdir( $install_dir, 'lib' ) );
                 LIB   => $install_dir );
}
my @scripts = qw( synctree.pl patchtree.pl mailrpt.pl W32Configure.pl
                  configsmoke.pl smokeperl.pl runsmoke.pl
                  archiverpt.pl smokestatus.pl chkbcfg.pl );
-f catfile( $findbin, '.patch' ) and push @scripts, '.patch';

WriteMakefile(
    NAME	  => "Test-Smoke",
    DISTNAME	  => "Test-Smoke",
    VERSION_FROM  => "lib/Test/Smoke.pm",
    $] >= 5.005 ? (
    AUTHOR	  => 'H.Merijn Brand <h.m.brand@hccnet.nl>, '.
		     'Nicholas Clark <nick@plum.flirble.org> and ' .
                     'Abe Timmerman <abeltje@cpan.org>',
    ABSTRACT_FROM => 'lib/Test/Smoke.pm',
    ) : (),

    dist	  => { COMPRESS => "gzip -9f",
                       SUFFIX   => ".gz",
		     },
    PREREQ_PM     => { 'File::Spec' => 0.82, 'Test::More' => 0},
    PMLIBDIRS     => [qw/ lib inc /],
    EXE_FILES     => \@scripts,
    %install
);

#sub MY::libscan {
#    my( $self, $path ) = @_;
#    my @nopaths = qw( .svn t private investigate );
#    my $patt = join '|', map {
#        /^\w/ and $_ = "\\b$_";
#        /\w$/ and $_ = "$_\\b";
#        $_;
#    } @nopaths;
#
#    return $path =~ m:$patt: ? "" : $path;
#}

sub dir_prompt {

    GETDIR: {
        my $dir = &prompt; # & -> pass @_

        # thanks to perlfaq5
        $dir =~ s{^ ~ ([^/]*)}
                 {$1 ? ( getpwnam $1 )[7] : 
                       ( $ENV{HOME} || $ENV{LOGDIR} || 
                         "$ENV{HOMEDRIVE}$ENV{HOMEPATH}" )}ex;

        my $cwd = cwd();
        my $dir_existed = -d $dir;
        File::Path::mkpath( $dir, 1, 0755 ) unless $dir_existed;
        chdir $dir or warn "Cannot chdir($dir): $!\n" and redo GETDIR;
        $dir = canonpath( cwd() );
        chdir $cwd or warn "Cannot chdir($cwd) back: $!\n";
        File::Path::rmtree( $dir ) unless $dir_existed;

        print "Got [$dir]\n";
        return $dir;
    }
}
